### ISSUE TITLE
駐車場検知

### ISSUE THREAD

**Issue by Hiromu1612** [Hiromu1612/Parking_space_detection_system_master#1]

**Comment by Hiromu1612** [comment]
# 6/16 9:30-12:00
今までの進捗状況のProjectsを間違えて消してしまった。。
macのm3proチップはtensorflow2系に対応できていないため、windowsでやったらすぐできた。

![Image](https://github.com/Hiromu1612/Parking_space_detection_system_master/assets/150511546/18b4de77-361e-4018-b690-1af500a26e49)

午後はKaggleの課題を進めた。Mask-rcnnでエラーが出てるから直す。

**Comment by Hiromu1612** [comment]
# 6/20 16:45-18:15, 20:30-1:45 (6.75h)
 最初にmacでtensorflowを動かすために、アーキテクチャを切り替えれるようにした。https://qiita.com/funatsufumiya/items/cec08f1ba3387edc2eed)
https://qiita.com/abemaru/items/1c7d0731c2179565a4fb
![image](https://github.com/Hiromu1612/Parking_space_detection_system_master/assets/150511546/879730f3-b330-42fb-8f8a-83104965a59e)
しかし、tensorflow1系は動作しなかった。そのため、以下のサイトを参考にtensorflow2系のDocker imageを使った。　

`FROM sonoisa/deep-learning-coding:pytorch1.12.0_tensorflow2.9.1`

https://data.gunosy.io/entry/tensorflow_on_m1

その結果、AMD64のエラーが消え、tensorflowをdocker上で実行できた。
![image](https://github.com/Hiromu1612/Parking_space_detection_system_master/assets/150511546/28e98bca-df78-4221-a72e-cc7c4a0eb8aa)

最後まで実行できたが、検知できていない。windowsのtensorflow1.15.2なら動くかもしれない。

![image](https://github.com/Hiromu1612/Parking_space_detection_system_master/assets/150511546/46865166-512d-4522-bd8b-a7bee0ec61d8)

**Comment by Hiromu1612** [comment]
# 11:00-12:00, 13:00-16:00, 22:00-1:00 (7h)
 13時に進捗会があった。matplotlibで描画ができていないのは、docker上で起動しているGUIアプリの画面と、ローカルのディスプレイを接続する必要があるからだった。

[https://qiita.com/Spritaro/items/f907a9b52cb78e4fbec0](url)　

このサイトは、mac用でターミナルで`xhost +local:root`を実行してから、`docker run -e DISPLAY=$DISPLAY -v /tmp/.X11-unix:/tmp/.X11-unix -it コンテナ名`で出来るが、windowsは以下のサイトのように2つのアプリをインストールしなければならない。

https://kamino.hatenablog.com/entry/docker_gui

https://mseeeen.msen.jp/windows-terminal-with-ubuntu/

https://www.u.tsukuba.ac.jp/ufaq/ssh%E6%8E%A5%E7%B6%9A%E7%B5%8C%E7%94%B1%E3%81%A7linux%E3%83%87%E3%82%B9%E3%82%AF%E3%83%88%E3%83%83%E3%83%97%E3%82%A2%E3%83%97%E3%83%AA%E3%82%B1%E3%83%BC%E3%82%B7%E3%83%A7%E3%83%B3%E3%82%92%E4%BD%BF/

具体的なコマンドの手順は
1. ターミナル　`docker build -t [イメージ名] .`
2. ターミナル 　`ssh hiromu@172.28.168.87`
3. Ubuntuのターミナル(勝手に開く)　`docker run -it -e DISPLAY=$DISPLAY -v /tmp/.X11-unix:/tmp/.X11-unix  [イメージ名] `
4. Ubuntuのターミナル　`docker exec -it [コンテナ名] /bin/bash`
5. Ubuntuのターミナル(Dockerコンテナ内)　　`python set_regions.py Parking.mp4` ←ここで処理が行われない

**Comment by Hiromu1612** [comment]
# 8/8(金) 10:00-12:00
Mask-RCNNをYOLOv7に変更。`detector.py`のmrcnnの読み込みをするコードを見ながら修正したが、まだまだ修正が必要。
Mask-RCNNの入出力データを理解することから始めると良いと教えてもらった。


![Image](https://github.com/user-attachments/assets/dfb48784-51f9-4bd6-9976-a0bc405517d6)

**Comment by Hiromu1612** [comment]
# 8/10(日) 20:00-23:00
一旦修正してみる。

`detector.py`の入力データは`rgb_image`でOpen-CVのBGRをRGBに変換している

それを踏まえて以下の二点を修正

![image](https://github.com/user-attachments/assets/60ddc0a4-489d-4475-80db-25a092ee5429)

![image](https://github.com/user-attachments/assets/11259a86-35fd-425b-ab49-03908480030a)

**Comment by Hiromu1612** [comment]
# 8/14(火) 22:00-24:40
`google_utils.py`の`attempt_download`関数のtag_listを修正

![image](https://github.com/user-attachments/assets/99c717a5-8355-46eb-b982-87c5f239ce62)

`detector.py`モデルのパスを変更

![image](https://github.com/user-attachments/assets/54a57895-3d03-4de4-9bb3-19893833242d)

`mc_bot_sort.py`のBoTSORTクラスの引数を設定

![image](https://github.com/user-attachments/assets/833ad251-8ac7-4754-aac0-4d2ee3b64d98)

argsを設定

![image](https://github.com/user-attachments/assets/c85b6171-f7ba-471c-a5b6-4924bd3f5e1e)

以下の2つのパスを自分で探した
![image](https://github.com/user-attachments/assets/810d50f4-62c6-4bdd-b715-f6ac9678650a)

**Comment by Hiromu1612** [comment]
# 8/15(金) 20:00-24:00
Dockerの環境構築でGPUではなくCPUで動作するようにすると良いと教えて頂いた。
`faiss-gpu`はエラーなくインストールできたが、`faiss-cpu`がインストールできないのが原因かもしれない。copilotを使いながらdockerfileを何度も編集したが、以下のエラーが出続けた。


![Image](https://github.com/user-attachments/assets/9d2afeb2-6786-49e6-a3d8-f1e6259b66e8)


解決策として考えられるのは
エラーなくインストールできた`faiss-gpu`をグラボ搭載PC環境で実行する 
　or 
ネットでさらに詳しく調べる

**Comment by Hiromu1612** [comment]
# 8/19 11:00-12:00, 13:00-16:00
Dockerfileで`faiss-cpu`がインストールできず、`swig`,`cmake`の依存関係が原因かと考えたがつまづいた。

質問したところ、そもそもDockerfileのベースイメージが今回のモデルはYOLOであるためtensorflowではなく、pytorchだった。

yolov7フォルダの`requirements.txt`のpytorchのバージョンをベースイメージに変更する。

![Image](https://github.com/user-attachments/assets/dfc68ae9-fcb2-4fa5-8463-7b4f57355f15)

**Comment by Hiromu1612** [comment]
#8/21(木) 
13時から進捗報告会があった。pythorchのGPUなしのベースイメージがなかったがあるはずと教えて頂いた。
メモ
・パレートの2.6.2の法則-- どんな集団も成果を出す2割、普通6割、足を引っ張る2割で構成される。
・前提・なぜそう考えたのか・試してダメだった理由をしっかりと起承転結の形で質問する(エラー文だけでは前提・試したことなどが分からず、答えを求められているように感じてしまう)
・結果をすぐに求めず、時間がかかっても良いから論理的思考力を養いながら進める(焦って進めても何も残らない)
・ChatGPTはあくまでも補助。web検索の方がより正確。コードの意味を順を追って理解しなければできない。今までの個人開発は簡単な内容でGPTでほとんど解決していたが、知識が身につかない

**Comment by Hiromu1612** [comment]
一旦MaskRCNNが動いていたコードに戻して考えることにした。
pytorch1.7.0のイメージを探したところ、
https://hub.docker.com/r/bitnami/pytorch/tags?page=22&page_size=&name=1.%20&ordering=
が見つかったので、Dockerfileの一行目を`FROM bitnami/pytorch:1.7.0`とした。

他にも、Pytorchの公式ドキュメントにCPUのみのバージョンがあった。
https://pytorch.org/get-started/previous-versions/

以下のような、`install pyproject.toml-based [projects`のエラーが出たが、[このサイト](https://postgresweb.com/python-error-which-is-required-to-install-pyproject-toml-based-projects)をもとに、`Microsoft Visual C++ Build Tools`をインストールした。

![Image](https://github.com/user-attachments/assets/bb9cf9f0-1d11-4b2c-8056-45498315f6d8)
